content/articles/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/index.md

```diff
@@ -0,0 +1,673 @@
+
+#### Client Dockerfile
+
+Navigate to the client folder and create a new file and name it `Dockerfile`. We will add the following to this file.
+
+```bash
+FROM node:alpine
+WORKDIR /app
+COPY package.json ./
+COPY package-lock.json ./
+COPY ./ ./
+RUN npm i
+CMD ["npm", "run", "start"]
+```
+
+The application will now be running on Docker. So we need to pull the Node.js image to run this application inside a container that will then be used to create our client image to run the React application. Inside the client image, we have defined the working directory as `/app`. When the image is created, this directory will also be created inside Docker. Then we will copy the `package-lock.json` and `package.json` from the computer to `/app` on the client's Docker image.
+
+After that, we will copy everything else from our client root folder to the client container's working directory. Then call the `npm runs start` command to start up React application in the container.
+
+Let's now test if this Dockerfile is working. On your terminal change directory to point to the client directory and execute this command;
+
+```bash
+docker build -f Dockerfile -t client .
+```
+
+The above command will create an image called client on Docker. To test it, we will create a container that will help us to run this image. Here is the command to do so.
+
+```bash
+docker run -it -p 4001:3000 client
+```
+
+Here will are exposing port 3000 (the port that runs the React application) to port 4001 outside the container—run the above command. And if now you open `http://localhost:4001` on a browser, you can see that the React app is working inside Docker.
+
+![dockerized-react-application](/engineering-education/build-and-dockerize-a-full-stack-react-app-with-nodejs-and-nginx/dockerized-react-application.png)
+
+#### Server Dockerfile
+
+```bash
+FROM node:alpine
+WORKDIR /app
+COPY package.json ./
+COPY package-lock.json ./
+COPY ./ ./
+RUN npm i
+CMD ["npm", "run", "start"]
+```
+
+This looks similar to the client Dockerfile, and the same process is being carried out here to create `/app` working directory, copy all the necessary files, and then run the `npm runs start` script (as defined in the package.json) to start up the server.
+
+We can test the Dockerfile by creating a server image and a container to run the image. To create the server image run;
+
+```bash
+docker build -f Dockerfile -t server .
+```
+
+To create a container to execute the server image, run;
+
+```bash
+docker run -it -p 4002:3001 server
+```
+
+Here will are exposing port 3001 (the port that is running the Node.js server application) to port 4002 outside the container—run the above command. And if now you open `http://localhost:4002/` on a browser, you can see that the Node.js home route is okay and the app is working inside Docker.
+
+![dockerized-nodejs-server](dockerized-nodejs-server.png)
+
+### Setting the Nginx server
```

Comment:
What is proxy? What is reverse proxy? Why do we use Nginx? What other alterative solutions do we have?